llvm/docs/SPIR-V-Backend.rst

```diff
@@ -0,0 +1,294 @@
+
+-  SPIRVInstrInfo
+
+   This class, aided by TableGen, defines and describes SPIR-V OpCodes.
+   This is currently hand-crafted with most SPIR-V OpCodes but could be
+   further scripted. See SPIRVInstrInfo.(h\|cpp\|td) for more details.
+
+-  SPIRVRegisterInfo and SPIRVRegisterBankInfo
+
+   These two classes are used to define information about registers for
+   SPIR-V. For SPIR-V, we only need very few register banks and classes.
+   Actually, we only need two: one for types and one for identifiers.
+
+-  SPIRVTypeRegistry
+
+   SPIRVTypeRegistry is used to maintain rich type information required
+   for SPIR-V even after lowering from LLVM IR to GMIR. It can convert
+   an llvm::Type into an OpTypeXXX instruction, and map it to a virtual
+   register using and ASSIGN\_TYPE pseudo instruction.
+
+   Type info from this class can only be used before it gets stripped
+   out by the InstructionSelector stage. All type info is function-local
+   until the final SPIRVGlobalTypesAndRegNums pass hoists it globally
+   and deduplicates it all.
+
+-  SPIRVLegalizerInfo
+
+   This class defines which operations are valid for SPIR-V.
+
+Execution Pipeline
+~~~~~~~~~~~~~~~~~~
+
+The main functions being executed, in that order:
+
+-  SPIRVBasicBlockDominance::runOnFunction
+
+   Sort the Basic Blocks so that blocks appear before all blocks they
+   dominate, as to satisfy SPIR-V validation rules.
+
+-  SPIRVIRTranslator::runOnMachineFunction
+
+   SPIRVIRTranslator overrides GlobalISel's IRTranslator to customize
+   the default lowering behavior. This is mostly to stop aggregate
+   types like Structs from getting expanded into scalars, and to
+   maintain type information required for SPIR-V using the
+   SPIRVTypeRegistry before it gets discarded as LLVM IR is lowered to
+   GMIR.
+
+   SPIRVIRTranslator is relying on SPIRVCallLowering to lower LLVM calls
+   into machine code calls. It relies on generateOpenCLBuiltinCall
+   (SPIRVOpenCLBIFs.cpp) for anything specific to OpenCL. Similarly,
+   SPIRVTypeRegistry will rely on generateOpenCLOpaqueType (in that same
+   file) for any type specific to OpenCL.
+
+-  Legalizer::runOnMachineFunction
+
+   This SPIR-V backend relies on the default Legalizer and
+   SPIRVLegalizerInfo.
+
+-  SPIRVInstructionSelect::runOnMachineFunction
+
+   SPIRVInstructionSelect is a custom subclass of InstructionSelect,
+   which is mostly the same except from not requiring RegBankSelect to
+   occur previously, and making sure a SPIRVTypeRegistry is initialized
+   before and reset after each run.
+
+   SPIRVInstructionSelect is helped by SPIRVInstructionSelector to
+   generate SPIR-V instructions from target-independent instructions.
+
+-  SPIRVBlockLabeler::runOnMachineFunction
+
+   SPIRVBlockLabeler ensures all basic blocks start with an OpLabel, and
+   ends with a suitable terminator.
+
+-  SPIRVAddRequirements::runOnMachineFunction
+
+   SPIRVAddRequirements iterates over all instructions and inserts
```

Comment:
The first part should already be implemented, but I agree the doc/comments should be updated to clarify that point. The configurability-part is currently listed as a TODO. I'll prepare a commit to update the doc soonish, and also add a test case.

Mind the fact that currently only `SPV_KHR_no_integer_wrap_decoration` is enabled: 
```cpp
// TODO use command line args for this rather than defaults
void SPIRVSubtarget::initAvailableExtensions(const Triple &TT) {
  using namespace Extension;
  if (TT.isVulkanEnvironment()) {
    availableExtensions = {};
  } else {
    // A default extension for testing - should use command line args
    availableExtensions = {SPV_KHR_no_integer_wrap_decoration};
  }
}
```
https://github.com/KhronosGroup/LLVM-SPIRV-Backend/pull/1/files#diff-9ac8724b38f0f2f16078e93ad5448a6fR151

Reply:
Thanks! BTW, what about restricting target SPIR-V version? I.e. is there any infrastructure to generate, for example, SPIR-V 1.2 compatible output file?

Reply:
We do have min/max version information about capabilities, decorations, etc... [1] that is used to ensure the output is valid, but currently we always assume SPIR-V 1.4 and there's a TODO to have a command line option, or something, to chose the desired SPIR-V version. [2] Once this TODO is implemented we should also add the expected tests, of course.

1. https://github.com/KhronosGroup/LLVM-SPIRV-Backend/pull/1/files#diff-3253aa7cd0b28412077e3f870178d895R105-R129
2. https://github.com/KhronosGroup/LLVM-SPIRV-Backend/pull/1/files#diff-9ac8724b38f0f2f16078e93ad5448a6fR45-R47

Reply:
Thanks!